core/logger/logger.go

```diff
@@ -173,6 +177,21 @@ func (l *zapLogger) ErrorIfClosing(c io.Closer, name string) {
 	}
 }
 
+func (l *zapLogger) Sql(query string, args ...interface{}) {
+	if l.logSql != nil && !l.logSql() {
+		return
+	}
+	if args != nil {
+		var pairs []string
+		for i, arg := range args {
+			pairs = append(pairs, fmt.Sprintf("$%d", i+1), fmt.Sprintf("%v", arg))
+		}
+		replacer := strings.NewReplacer(pairs...)
+		query = replacer.Replace(query)
+	}
+	l.Debugf("SQL: %s", query)
+}
```

Comment:
Is there a compelling reason to have this SQL specific code inside of the loggers themselves? I had pictured all of this happening a layer above, inside of `Q`.

Reply:
If you package up this string formatting logic in a `String()` method on a helper struct, then you could pass a `queryFmt{query, args}` so that the computation is deferred until/unless it's actually printed.

Reply:
I addressed this request, but I have one concern: `q` needs two things:
- a Logger
- a config that holds LogSqlStatements value
Currently, I have Logger persisting this setting, but internally Logger does not use it and I found this odd.
Ideally for every NewQ we need to pass: a Logger and a Config, but this seems little overkill, too much parameters to construct a Q.
Can we please discuss this: how do you see Q accessing `LogSqlStatements` from Config?

Reply:
Discussed in Slack. To be updated in the next iteration.